### PR TITLE
Issue #811 Merge DefaultStoreProvider into CacheStore.Provider

### PR DESCRIPTION
**Pull request by cljohnso**
This commit merges the tiering logic from DefaultStoreProvider into
CacheStore.Provider and discards DefaultStoreProvider.

### CODE REVIEW

File: impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java [hunk 1]
```diff
@@ -103,6 +110,8 @@ public void onInvalidation(K key, ValueHolder<V> valueHolder) {
   }
 
   static class ComputationException extends RuntimeException {
+    private static final long serialVersionUID = -7117636543507991092L;
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java]: Unrelated... if we want to 'fix' this we should do it in one fell swoop.

File: impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java [hunk 2]
```diff
@@ -336,48 +345,126 @@ public void clear() throws CacheAccessException {
...
 
+  @ServiceDependencies({CompoundCachingTier.Provider.class,
+      OnHeapStore.Provider.class, OffHeapStore.Provider.class, OffHeapDiskStore.Provider.class})
   public static class Provider implements Store.Provider {
 
+    private static final Set<Set<ResourceType.Core>> SUPPORTED_RESOURCE_COMBINATIONS;
+    static {
+      // Logic in setTierConfigurations must mirror this set
+      final Set<Set<ResourceType.Core>> supported = new HashSet<Set<ResourceType.Core>>();
+      supported.add(EnumSet.of(ResourceType.Core.HEAP, ResourceType.Core.DISK));
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java]: static imports would make this a lot cleaner.

File: impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java [hunk 3]
```diff
@@ -336,48 +345,126 @@ public void clear() throws CacheAccessException {
...
+      supported.add(EnumSet.of(ResourceType.Core.HEAP, ResourceType.Core.OFFHEAP));
+      supported.add(EnumSet.of(ResourceType.Core.HEAP, ResourceType.Core.OFFHEAP, ResourceType.Core.DISK));
+      SUPPORTED_RESOURCE_COMBINATIONS = supported;
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java]: unmodifiableSet(...) would be a sensible defensive move here.

File: impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java [hunk 4]
```diff
@@ -336,48 +345,126 @@ public void clear() throws CacheAccessException {
...
+        return 0;
+      }
     }
 
     @Override
     public <K, V> Store<K, V> createStore(Configuration<K, V> storeConfig, ServiceConfiguration<?>... serviceConfigs) {
       CacheStoreServiceConfiguration cacheStoreServiceConfig = findSingletonAmongst(CacheStoreServiceConfiguration.class, (Object[])serviceConfigs);
-      if (cacheStoreServiceConfig == null) {
-        throw new IllegalArgumentException("Cache store cannot be configured without explicit config");
+      if (cacheStoreServiceConfig != null) {
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java]: If I'm reading whats going on here right, would it maybe be simpler to make CacheStoreServiceConfiguration not implement ServiceConfiguration, and then rename it to CacheStoreConfiguration?

File: impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java (hunk 3 above)
Comment (cljohnso) [inline impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java]: You'd do that when the field value is used only internally?

File: impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java (hunk 4 above)
Comment (cljohnso) [inline impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java]: I really have no clue what you're talking about here; care to elaborate?

File: impl/src/test/java/org/ehcache/impl/internal/store/tiering/CacheStoreTest.java [hunk 5]
```diff
@@ -63,28 +71,36 @@
...
 public class CacheStoreTest {
 
+  @Mock
+  private CachingTier<Number, CharSequence> numberCachingTier;
+  @Mock
+  private AuthoritativeTier<Number, CharSequence> numberAuthoritativeTier;
+  @Mock
+  private CachingTier<String, String> stringCachingTier;
+  @Mock
+  private AuthoritativeTier<String, String> stringAuthoritativeTier;
```
Comment (chrisdennis) [inline impl/src/test/java/org/ehcache/impl/internal/store/tiering/CacheStoreTest.java]: If this is just eliminating doing the mocking in the first two lines of each test, then consider my vote a -1 against this specific change... just a vote however.  Would like others to weigh in.

File: impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java (hunk 2 above)
Comment (cljohnso) [inline impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java]: You have two levels of static import that could be used here -- at what level would you recommend?

File: impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java (hunk 3 above)
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java]: If you are against I can live without it.

File: impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java (hunk 2 above)
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java]: `supported.add(of(HEAP, DISK));`

File: impl/src/test/java/org/ehcache/impl/internal/store/tiering/CacheStoreTest.java (hunk 5 above)
Comment (cljohnso) [inline impl/src/test/java/org/ehcache/impl/internal/store/tiering/CacheStoreTest.java]: This is actually the recommended way of mocking classes with generic parameters ...

File: impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java (hunk 4 above)
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java]: Agreed with Chris + making config class package private

Comment (cljohnso) [inline impl/src/main/java/org/ehcache/impl/internal/store/tiering/CacheStore.java]: Just got out of a hangout with Chris -- we're going to reduce the `CacheStoreServiceConfiguration` to it's minimal form and make it an internal class.  There are a couple of tests that would need adjustment but that configuration is really only used internally to `CacheStore.Provider`.

**Comment by cljohnso** [comment]
Updated and rebased ...

**Pull request merged**